src/Native/System.Native/pal_mount.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_mount.h"
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+
+// Check if we should use getmntinfo or /proc/mounts
+#if HAVE_MNTINFO
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#include <mntent.h>
+#define STRING_BUFFER_SIZE 8192
+#endif
+
+#if HAVE_MNTINFO
+
+int GetMountInfo(MountPointFound onFound)
+{
+    // getmntinfo returns pointers to OS-internal structs, so we don't need to worry about free'ing the object
+    int result = 0;
+    struct statfs* mounts = nullptr;
+    int count = getmntinfo(&mounts, 0);
+    if (count > 0)
+    {
+        for (int32_t i = 0; i < count; i++)
+        {
+            int32_t nameLength = strlen(mounts[i].f_mntonname);
+            onFound(mounts[i].f_mntonname, nameLength);
+        }
+    }
+    else
+    {
+        // Something went wrong so set our result to errno
+        result = errno;
+    }
+    
+    return result;
+}
+
+#else
+
+int GetMountInfo(MountPointFound onFound)
+{
+    int result = 0;
+    FILE* fp = setmntent("/proc/mounts", MNTOPT_RO);
+    if (fp == nullptr)
+    {
+        result = errno;
+    }
+    else
+    {
+        // The _r version of getmntent needs all buffers to be passed in; however, we don't know how big of a string
+        // buffer we will need, so pick something that seems like it will be big enough.
+        char buffer[STRING_BUFFER_SIZE] = { };
+        mntent entry = { };
+        while (getmntent_r(fp, &entry, buffer, STRING_BUFFER_SIZE) != nullptr)
+        {
+            int32_t nameLength = strlen(entry.mnt_dir);
+            onFound(entry.mnt_dir, nameLength);
+            memset(buffer, 0, STRING_BUFFER_SIZE);
+        }
+
+        assert(endmntent(fp) == 1); // documented to always return 1
+        result = 0; // We need to standardize a success return code between our implementations, so settle on 0 for success
+    }
+
+    return result;
+}
+
+#endif
+
+extern "C"
+int GetAllMountPoints(MountPointFound onFound)
+{
+    return GetMountInfo(onFound);
+}
+
+extern "C"
+int GetSpaceInfoForMountPoint(
+    const char*             name,
+    MountPointInformation*  mpi)
+{
+    assert(name != nullptr);
+    assert(mpi != nullptr);
+
+    struct statfs stats = { };
+    int result = statfs(name, &stats);
+    if (result == 0)
+    {
+        mpi->AvailableFreeSpace = (int64_t)(stats.f_bsize * stats.f_bavail);
+        mpi->TotalFreeSpace = (int64_t)(stats.f_bsize * stats.f_bfree);
+        mpi->TotalSize = (int64_t)(stats.f_bsize * stats.f_blocks);
+    }
+    else
+    {
+        result = errno;
```

Comment:
The P/Invoke for GetSpaceInfoForMountPoint is written using ```out``` for the mpi argument.  That means the C# compiler will expect ```mpi``` will have been set by the time the function returns.  We should be zero'ing it out in the error path.